Add all definitions of UVM-RAL classes

RasmusGOlsen opened the pull request:
This commit marks the beginning of the cleanup process and updates the UVM Register Layer to align with IEEE 1800.2-2020 standards. I have created a new directory called **reg** where all UVM Register Layer files are now located. The structure follows the original SystemVerilog implementation found [here](https://github.com/accellera-official/uvm-core) to facilitate tracking progress and comparisons. 

Below is a list of tasks that will help with the transition.

- [x] Define all uvm_reg* classes
- [x] Define methods for all uvm_reg* classes (in progress)
- [x] Migrate existing implementation
    - [x] uvm_reg_field
    - [x] uvm_reg
    - [x] uvm_mem
    - [x] uvm_reg_block
    - [x] uvm_reg_map
    - [x] uvm_reg_item
    - [x] uvm_reg_adapter
    - [x] uvm_reg_predict
- [x] Create test cases for each class

Resolves: #226 , Resolves: #227

ktbarrett commented:
@RasmusGOlsen You can see the issue by clicking on the link:
```
ruff check...............................................................Passed
ruff format..............................................................Failed
- hook id: ruff-format
- files were modified by this hook

23 files reformatted, 61 files left unchanged

trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
fix utf-8 byte order marker..............................................Passed
check yaml...............................................................Passed
check toml...............................................................Passed
check for added large files..............................................Passed
codespell................................................................Passed
```

Looks like your files need formatting changes.

The best way to accomplish that (and the best way to develop pyuvm) is to follow the [Contributing docs](https://github.com/pyuvm/pyuvm?tab=readme-ov-file#contributing) which go over how to install pre-commit and use it.

raysalemi commented:
Okay. So let's go forward with hints in RAL and @ktbarrett, we can give up support for 3.6 if **cocotb** is doing that.

I'm afraid there won't be a second edition of the book.  Also, I would hate to have people who bought the current book own an obsolete book.  

The compatibility issue goes beyond the book, as there are hundreds of people who have written code in the style of the book, and we don't want to break them.

cmarqu commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,935 @@
+                self.build()
+            except AttributeError as e:
+                logger.error(
+                    "Calling build() when not implemented by user"
```
```suggestion
                    "Calling build() when not implemented by user\n"
```

cmarqu commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,935 @@
+        if len(self._maps) > 1 and not map:
+            logger.error(
+                f"'set_offset' requires a map when register "
+                f"'{self.get_full_name()}' belongs to more than one map."
```
```suggestion
                f"{self.get_full_name()!r} belongs to more than one map."
```
Using `repr()` instead of the default `str()`.

cmarqu commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,935 @@
+        if not isinstance(path, uvm_door_e):
+            warnings.warn(
+                "The 'write' arguments has changed in order to "
+                "match IEEE 1800.2 standard.",
```
```suggestion
                "match the IEEE 1800.2 standard.",
```

cmarqu commented on pyuvm/reg/uvm_reg_map.py:
```suggestion
                    f"uvm_reg_item 'element_kind' is UVM_FIELD but 'element' "
```

cmarqu commented on tests/pytests/reg/test_uvm_reg_block.py:
```diff
@@ -0,0 +1,625 @@
+If pip install pytest-sugar is ran then pytest is gonna likely execute a
+bar progression while
+running tests (especially if in Parallel)
+"""
```
IMO a bit strange here, how about moving it into a README file?

RasmusGOlsen commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,935 @@
+            else:
+                map_name = map.get_full_name()
+            logger.warning(
+                f"Register '{self.get_name()}' is unmapped in map '{map_name}'"
```
resolved using `repr()`

ktbarrett commented on tests/cocotb_tests/test_ral_read_write/Makefile:
This isn't supported in cocotb < 2.0. This is likely breaking these tests. Leave it be.

ktbarrett commented on tests/cocotb_tests/test_ral_read_write/test.py:
```diff
@@ -51,38 +67,42 @@ def __init__(self, name):
-        self.CMD_data = BinaryValue(0, 32, self.Endian)
-        self.WR_RD = BinaryValue(0, 1, self.Endian)
+        self.CMD_data = LogicArray(0, 32)
+        self.WR_RD = LogicArray(0, 1)
```
This constructor isn't supported in cocotb < 2.0. The only reason this isn't breaking is because the previous comment is causing no tests to be discovered, which before 2.0 was just a warning.

ktbarrett commented on tests/cocotb_tests/test_ral_read_write/test.py:
```diff
@@ -51,38 +67,42 @@ def __init__(self, name):
         self.write_data: int = 0
         self.read_data: int = 0
-        self.addr: str = ""
+        self.addr: int = None
```
Either remove type annotations or use them correctly and check them. We would avoid implicit optionals.

ktbarrett reviewed (commented):
I'll come back to it once the current issues are addressed.

RasmusGOlsen commented on tests/cocotb_tests/test_ral_read_write/test.py:
```diff
@@ -153,41 +171,41 @@ def __init__(self, name="cmd_adapter"):
-        if rw.kind == access_e.UVM_READ:
-            cmdItem.WR_RD.integer = 0
+        if rw.kind == uvm_access_e.UVM_READ:
+            cmdItem.WR_RD[:] = 0
```
This is now an `int`.

ktbarrett commented on tests/pytests/reg/test_uvm_ral.py:
```diff
@@ -0,0 +1,146 @@
+##############################################################################
+
+
+@pytest.fixture
```
This isn't an appropriate use of fixture.

ktbarrett commented on pyuvm/reg/uvm_mem.py:
```diff
@@ -0,0 +1,486 @@
+
+
+class uvm_mem(uvm_object):
+    _max_size: int = 0
```
```suggestion
    _max_size: ClassVar[int] = 0
```

ktbarrett commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,931 @@
+    def reset(self, kind: str = "HARD") -> None:
+        for field in self._fields:
+            field.reset(kind)
+        self._atomic.release()
```
I don't see a corresponding acquire? This will raise an exception when using a `Lock` object.

RasmusGOlsen replied:
I will add a test if it is acquired or do a try-except.

RasmusGOlsen commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,931 @@
+
+import logging
+import warnings
+from asyncio import Semaphore
```
I will take a look at this. There are also no unit tests that cover this feature.

ktbarrett commented on pyuvm/reg/uvm_mem.py:
```diff
@@ -0,0 +1,486 @@
+    from pyuvm.s14_15_python_sequences import uvm_sequence_base
+
+__all__ = ["uvm_mem"]
+logger = logging.getLogger("RegModel")
```
Does UVM have the concept of a canonical name for components in the hierarchy? It might be a good idea to have each instance have its own logger based on that name.

Feel free to ignore this for now. This can be follow-on.

ktbarrett commented on pyuvm/reg/uvm_reg.py:
```diff
@@ -0,0 +1,939 @@
+if os.getenv("COCOTB_TEST_MODULES") is None and os.getenv("MODULE") is None:
+    from cocotb.triggers import Lock
+else:
+    from asyncio import Semaphore as Lock
```
Ooof. I'm not sure this is the best way to handle this. Is this for production, or just for testing? If it's just for testing it's probably fine for now, but we need a story if this is for production use. Also use `asyncio.Lock`, Semaphore is the wrong thing here.

RasmusGOlsen replied:
The idea is to use the tests for hardware production test. This should be achived by writting an appropiate adapter class. We can also wait with this, since the uvm_sequence_item also needs to be updated to support this. I can make a new PR when I have made a proff of concept to see if it works. One question though. Are there a better way to test if we are running Cocotb?

ktbarrett commented on tests/pytests/reg/compatibility/test_compat_uvm_ral.py:
```diff
@@ -32,6 +37,7 @@
 ##############################################################################
 
 
+@pytest.mark.skip(reason="Using str as address is not backward compatible")
```
This was not addressed. Provide rationale or fix it.

RasmusGOlsen commented on pyuvm/reg/uvm_reg_field.py:
```diff
@@ -0,0 +1,666 @@
+logger = logging.getLogger("RegModel")
+
+
+_PREDEFINED_POLICIES: set[str] = set(
```
No it should not be an enum, because need to be able to add user-defined policies.

ktbarrett commented:
@RasmusGOlsen Let me know when this is ready for a final review.

RasmusGOlsen commented:
I think the only outstanding issue is with the Sphinx documentation is not compatible with typehints.

RasmusGOlsen commented:
I have now cleaned up the old implementation and updated the documentation build flow so that it handles type checking correctly. If there are any more issues, please let me know.

ktbarrett commented on tests/pytests/reg/test_uvm_reg_model.py:
```diff
@@ -5,6 +5,7 @@
 pytestmark = pytest.mark.usefixtures("initialize_pyuvm")
 
 
+@pytest.mark.skip(reason="Not implemented")
```
These should definitely be
```suggestion
@pytest.mark.xfail(raises=NotImplementedError)
```

This will ensure if the functionality is added in the future this test will start failing, signaling it should have the mark removed and the test re-evaluated for correctness.

ktbarrett commented:
Created #302.
